database.rules.json

```diff
@@ -9,7 +9,7 @@
         },
         "endedAt": {
           ".write": "auth != null && data.parent().child('users/' + auth.uid).exists() && !data.exists() && newData.exists()",
```

Comment:
You can get rid of lines 10-13. The `endedAt` field does not need to be written from an external user, just from the trusted serverless admin API, which has full access anyway.